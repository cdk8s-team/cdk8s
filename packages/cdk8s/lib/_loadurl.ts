// this is an executable program that downloads a file from a url (with
// redirects) and prints it's output to STDOUT. it is implemented as a program
// because node.js does not support issuing network requests synchronously and
// in the cdk we need everything to be synchronous (it all happens in the
// ctors). so we utilize `spawnSync` to spawn this program as a child process.
// alternatively we could have use `curl` but this is more portable.

import got from 'got';
import { parse } from 'url';
import * as fs from 'fs';

async function load(url: string) {
  const purl = parse(url);

  // read from filesystem if this is a file
  if (!purl.protocol || purl.protocol === 'file:') {
    if (!purl.pathname) {
      throw new Error(`unable to parse pathname from file url: ${url}`);
    }

    return fs.readFileSync(purl.pathname, 'utf-8');
  }

  const response = await got(url, {
    followRedirect: true,
    throwHttpErrors: true,
  });

  return response.body;
}

load(process.argv[2])
  .then(out => process.stdout.write(out))
  .catch((e: Error) => {
    console.error(e.stack);
    process.exit(1);
  });
