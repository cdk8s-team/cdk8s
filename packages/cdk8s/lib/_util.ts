export function removeEmpty(obj: any): any {
  if (obj == null) {
    return undefined;
  }

  if (typeof(obj) !== 'object') {
    return obj;
  }

  if (Array.isArray(obj)) {
    return obj.map(x => removeEmpty(x));
  }

  if (obj.constructor.name !== 'Object') {
    throw new Error(`can't render non-simple object of type '${obj.constructor.name}'`);
  }

  const newObj: { [key: string]: any } = { };

  for (const [key, value] of Object.entries(obj)) {
    const newValue = removeEmpty(value);
    if (newValue != null) {
      newObj[key] = newValue;
    }
  }

  return newObj;
}