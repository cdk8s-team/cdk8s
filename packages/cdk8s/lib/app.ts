import { Construct, ConstructNode } from '@aws-cdk/core';
import * as fs from 'fs';
import * as path from 'path';

export interface AppOptions {
  /**
   * The directory to output Kubernetes manifests.
   * 
   * @default "dist"
   */
  readonly outdir?: string;
}

/**
 * Represents a cdk8s application.
 */
export class App extends Construct {
  /**
   * The output directory into which manifests will be synthesized.
   */
  public readonly outdir: string;

  /**
   * Defines an app
   * @param options configuration options
   */
  constructor(options: AppOptions = { }) {
    super(undefined as any, '');
    this.outdir = options.outdir ?? 'dist';
  }

  /**
   * Synthesizes all manifests to the output directory
   */
  public synth(): void {
    ConstructNode.synth(this.node, {
      outdir: this.outdir
    });

    // remove manifest.json and cdk.out
    rm(path.join(this.outdir, 'manifest.json'));
    rm(path.join(this.outdir, 'cdk.out'));
  }
}

function rm(filePath: string) {
  if (fs.existsSync(filePath)) {
    fs.unlinkSync(filePath);
  }
}