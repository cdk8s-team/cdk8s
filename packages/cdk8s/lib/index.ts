export * from './api-object';
export * from './chart';
export * from './testing';
export * from './app';
