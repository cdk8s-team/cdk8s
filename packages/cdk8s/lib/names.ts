import * as crypto from 'crypto';

const MAX_DNS_NAME_LEN = 63;
const VALIDATE = /^[0-9a-z-]+$/;
const HASH_LEN = 8;

/**
 * Utilities for generating unique and stable names.
 */
export class Names {
  /**
   * Generates a unique and stable name compatible DNS_LABEL from RFC-1123 from
   * a path.
   *
   * The generated name will:
   *  - contain at most 63 characters
   *  - contain only lowercase alphanumeric characters or ‘-’
   *  - start with an alphanumeric character
   *  - end with an alphanumeric character
   *
   * The generated name will have the form:
   *  <comp0>-<comp1>-..-<compN>-<short-hash>
   *
   * Where <comp> are the path components (assuming they are is separated by
   * "/").
   *
   * Note that if the total length is longer than 63 characters, we will trim
   * the first components since the last components usually encode more meaning.
   *
   * @link https://tools.ietf.org/html/rfc1123
   *
   * @param path a path to a node (components separated by "/")
   * @param maxLen maximum allowed length for name
   * @throws if any of the components do not adhere to naming constraints or
   * length.
   */
  public static toDnsLabel(path: string, maxLen = MAX_DNS_NAME_LEN) {
    if (maxLen < HASH_LEN) {
      throw new Error(`minimum max length for object names is ${HASH_LEN} (required for hash)`);
    }

    const components = path.split('/');

    // verify components only use allowed chars.
    for (const comp of components) {
      if (!VALIDATE.test(comp)) {
        throw new Error(`"${comp}" is not a valid object name. The characters allowed in names are: digits (0-9), lower case letters (a-z), -, and .`);
      }

      if (comp.length > maxLen) {
        throw new Error(`Object name "${comp}" is too long. Maximum allowed length is ${maxLen}`);
      }
    }

    // special case: if we only have one component in our path, we don't decorate it
    if (components.length === 1) {
      return components[0];
    }

    components.push(calcHash(path, HASH_LEN));

    return components
      .reverse()
      .join('/')
      .slice(0, maxLen)
      .split('/')
      .reverse()
      .filter(x => x)
      .join('-');
  }

  /* istanbul ignore next */
  private constructor() {
    return;
  }
}

function calcHash(path: string, maxLen: number) {
  const hash = crypto.createHash('sha256');
  hash.update(path);
  return hash.digest('hex').slice(0, maxLen);
}