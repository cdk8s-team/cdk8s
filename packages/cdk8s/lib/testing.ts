import fs = require('fs');
import path = require('path');
import os = require('os');
import { App, Chart } from '../lib';

/**
 * Testing utilities for cdk8s applications.
 */
export class Testing {
  /**
   * Returns an app for testing with the following properties:
   * - Output directory is a temp dir.
   */
  public static app() {
    const outdir = fs.mkdtempSync(path.join(os.tmpdir(), 'cdk8s.outdir.'));
    return new App({ outdir });
  }

  /**
   * Returns the Kubernetes manifest synthesized from this chart.
   */
  public static synth(chart: Chart) {
    return chart.toJson();
  }

  /* istanbul ignore next */
  private constructor() {
    return;
  }
}
