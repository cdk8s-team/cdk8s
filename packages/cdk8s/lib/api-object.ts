import { Construct } from 'constructs';
import { Chart } from './chart';
import { removeEmpty } from './_util';
import { resolve } from './_tokens';
import * as stringify from 'json-stable-stringify';

/**
 * Metadata associated with this object.
 */
export interface ApiObjectMetadata {
  /**
   * The unique, namespace-global, name of this object inside the Kubernetes
   * cluster.
   *
   * Normally, you shouldn't specify names for objects and let the CDK generate
   * a name for you that is application-unique. The names CDK generates are
   * composed from the construct path components, separated by dots and a suffix
   * that is based on a hash of the entire path, to ensure uniqueness.
   *
   * You can supply custom name allocation logic by overriding the
   * `chart.generateObjectName` method.
   *
   * If you use an explicit name here, bear in mind that this reduces the
   * composability of your construct because it won't be possible to include
   * more than one instance in any app. Therefore it is highly recommended to
   * leave this unspecified.
   *
   * @default - an app-unique name generated by the chart
   */
  readonly name?: string;

  /**
   * Arbitrary key/value metadata.
   */
  readonly [key: string]: any;
}

/**
 * Options for defining API objects.
 */
export interface ApiObjectOptions {
  /**
   * Data associated with the resource.
   */
  readonly data?: any;

  /**
   * Object metadata.
   *
   * If `name` is not specified, an app-unique name will be allocated by the
   * framework based on the path of the construct within thes construct tree.
   */
  readonly metadata?: ApiObjectMetadata;  

  /**
   * API version.
   */
  readonly apiVersion: string;

  /**
   * Resource kind.
   */
  readonly kind: string;

  /**
   * Additional attributes for this API object.
   */
  readonly [key: string]: any;
}

export class ApiObject extends Construct {
  /**
   * The name of the API object.
   *
   * If a name is specified in `metadata.name` this will be the name returned.
   * Otherwise, a name will be generated by calling
   * `Chart.of(this).generatedObjectName(this)`, which by default uses the
   * construct path to generate a DNS-compatible name for the resource.
   */
  public readonly name: string;

  /**
   * The object's API version.
   */
  public readonly apiVersion: string;

  /**
   * The object kind.
   */
  public readonly kind: string;

  /**
   * The chart in which this object is defined.
   */
  public readonly chart: Chart;

  /**
   * Defines an API object.
   * 
   * @param scope the construct scope
   * @param ns namespace
   * @param options options
   */
  constructor(scope: Construct, ns: string, private readonly options: ApiObjectOptions) {
    super(scope, ns);
    this.chart = Chart.of(this);
    this.kind = options.kind;
    this.apiVersion = options.apiVersion;
    this.name = options.metadata?.name ?? this.chart.generateObjectName(this);
  }

  /**
   * Renders the object to Kubernetes JSON.
   */
  public toJson(): any {
    const data = {
      ...this.options,
      metadata: {
        namespace: Chart.of(this).namespace,
        ...this.options.metadata,
        name: this.name,
      }
    };

    // convert to "pure data" so, for example, when we convert to yaml these
    // references are not converted to anchors.
    return JSON.parse(stringify(removeEmpty(resolve(this, data))));
  }
}
