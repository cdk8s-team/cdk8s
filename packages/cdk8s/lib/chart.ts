import { Construct, ISynthesisSession } from '@aws-cdk/core';
import * as fs from 'fs';
import * as path from 'path';
import { ApiObject } from './api-object';
import * as YAML from 'yaml';
import { Names } from './names';

export class Chart extends Construct {

  /**
   * Finds the chart in which a node is defined.
   * @param node a construct node
   */
  public static of(node: Construct): Chart {
    if (node instanceof Chart) {
      return node;
    }

    const parent = node.node.scope as Construct;
    if (!parent) {
      throw new Error(`cannot find a parent chart (directly or indirectly)`);
    }

    return Chart.of(parent);
  }

  /**
   * The name of the stack's YAML file as emitted into the cloud assembly
   * directory during synthesis.
   */
  public readonly manifestFile: string;

  constructor(scope: Construct, ns: string) {
    super(scope, ns);
    this.manifestFile = `${this.node.uniqueId}.k8s.yaml`;
  }

  /**
   * Generates a app-unique name for an object given it's construct node path.
   *
   * Different resource types may have different constraints on names
   * (`metadata.name`). The previous version of the name generator was
   * compatible with DNS_SUBDOMAIN but not with DNS_LABEL.
   *
   * For example, `Deployment` names must comply with DNS_SUBDOMAIN while
   * `Service` names must comply with DNS_LABEL.
   *
   * Since there is no formal specification for this, the default name
   * generation scheme for kubernetes objects in cdk8s was changed to DNS_LABEL,
   * since itâ€™s the common denominator for all kubernetes resources
   * (supposedly).
   *
   * You can override this method if you wish to customize object names at the
   * chart level.
   *
   * @param apiObject The API object to generate a name for.
   */
  public generateObjectName(apiObject: ApiObject) {
    return Names.toDnsLabel(apiObject.node.path);
  }

  /**
   * Renders this chart to a set of Kubernetes JSON resources.
   * @returns array of resource manifests
   */
  public toJson(): any[] {
    return this.node.findAll().filter(x => x instanceof ApiObject).map(x => (x as ApiObject).toJson());
  }

  /**
   * Called by the app to synthesize the chart as a YAML file in the output directory/
   */
  protected synthesize(session: ISynthesisSession) {
    // convert each resource to yaml and separate with a '---' line
    const doc = this.toJson().map(r => YAML.stringify(r)).join('---\n');
    fs.writeFileSync(path.join(session.assembly.outdir, this.manifestFile), doc);
  }
}
