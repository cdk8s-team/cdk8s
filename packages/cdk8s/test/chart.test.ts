import { Chart, ApiObject, Testing } from '../lib';
import { Construct, Lazy } from '@aws-cdk/core';

test('empty stack', () => {
  // GIVEN
  const app = Testing.app();

  // WHEN
  const chart = new Chart(app, 'empty');

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
});

test('output includes all synthesized resources', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'test');

  // WHEN
  new ApiObject(chart, 'resource1', { kind: 'Resource1', apiVersion: 'v1' });
  new ApiObject(chart, 'resource2', { kind: 'Resource2', apiVersion: 'v1' });
  new ApiObject(chart, 'resource3', { kind: 'Resource3', apiVersion: 'v1' });

  // also, subscopes
  const scope = new Construct(chart, 'scope');
  new ApiObject(scope, 'resource1', { kind: 'Resource1', apiVersion: 'v1' });
  new ApiObject(scope, 'resource2', { kind: 'Resource2', apiVersion: 'v1' });

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
});


test('tokens are resolved during synth', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'test');

  // WHEN
  new ApiObject(chart, 'resource1', {
    kind: 'Resource1',
    apiVersion: 'v1',
    spec: {
      foo: Lazy.numberValue({ produce: () => 123 }),
      implicitToken: createImplictToken({ foo: 'bar' })
    }
  });

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
});

test('Chart.of(node) returns the first chart in which a node is defined', () => {
  // GIVEN
  const app = Testing.app();
  
  // WHEN
  const chart = new Chart(app, 'MyFirst');
  const direct = new Construct(chart, 'Direct');
  const indirect = new Construct(direct, 'Indirect');

  const childChart = new Chart(indirect, 'ChildChart');
  const childChild = new Construct(childChart, 'ChildChild');

  expect(Chart.of(chart)).toEqual(chart);
  expect(Chart.of(direct)).toEqual(chart);
  expect(Chart.of(indirect)).toEqual(chart);
  expect(Chart.of(childChart)).toEqual(childChart);
  expect(Chart.of(childChild)).toEqual(childChart);
});

test('Chart.of(node) fails when there is no chart in the tree', () => {
  // GIVEN
  const app = Testing.app();
  const child = new Construct(app, 'MyConstruct');

  // WHEN
  expect(() => Chart.of(child)).toThrow(/cannot find a parent chart \(directly or indirectly\)/);
});

test('synthesizeManifest() can be used to synthesize a specific chart', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'chart');
  new ApiObject(chart, 'obj1', { apiVersion: 'v1', kind: 'Kind1' });
  new ApiObject(chart, 'obj2', { apiVersion: 'v1', kind: 'Kind2' });

  // WHEN
  const manifest = chart.toJson();

  // THEN
  expect(manifest).toMatchSnapshot();
});

function createImplictToken(value: any) {
  const implicit = {};
  Object.defineProperty(implicit, 'resolve', { value: () => value });
  return implicit;
}