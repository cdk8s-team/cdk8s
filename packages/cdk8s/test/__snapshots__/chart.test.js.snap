// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty stack 1`] = `Array []`;

exports[`output includes all synthesized resources 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Resource1",
    "metadata": Object {
      "name": "test-resource1-2b59e3b8",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Resource2",
    "metadata": Object {
      "name": "test-resource2-22077398",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Resource3",
    "metadata": Object {
      "name": "test-resource3-365bba40",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Resource1",
    "metadata": Object {
      "name": "test-scope-resource1-3d54e80a",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Resource2",
    "metadata": Object {
      "name": "test-scope-resource2-e1b938d2",
    },
  },
]
`;

exports[`synthesizeManifest() can be used to synthesize a specific chart 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Kind1",
    "metadata": Object {
      "name": "chart-obj1-f2e469b6",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Kind2",
    "metadata": Object {
      "name": "chart-obj2-e10626fd",
    },
  },
]
`;

exports[`tokens are resolved during synth 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Resource1",
    "metadata": Object {
      "name": "test-resource1-2b59e3b8",
    },
    "spec": Object {
      "foo": 123,
      "implicitToken": Object {
        "foo": "bar",
      },
    },
  },
]
`;
